content/blog/2023/2023-11-20-design-patterns.md

```diff
@@ -0,0 +1,641 @@
+
+Maintainability:
+Design patterns enhance code maintainability by providing a clear and organized structure. When developers are familiar with common design patterns, it becomes easier for them to understand and modify the code, reducing the chances of introducing bugs during maintenance.
+
+Scalability:
+Design patterns contribute to the scalability of a codebase by providing modular and extensible solutions. As our application evolves, we can add new features or modify existing ones without having to overhaul the entire codebase.
+
+Abstraction and Encapsulation:
+Design patterns often involve abstraction and encapsulation, which help in hiding the complexity of the implementation details. This separation allows developers to focus on high-level design decisions without getting bogged down by low-level details.
+
+Flexibility:
+Design patterns make code more flexible and adaptable to change. When the structure of our software is based on well-established patterns, it becomes easier to introduce new functionality or modify existing behavior without affecting the entire system.
+
+Code Understandability:
+Design patterns provide a common vocabulary for developers. When a developer sees a particular pattern being used, they can quickly understand the intent and functionality without delving deeply into the implementation details.
+
+Testability:
+Code that follows design patterns is often more modular and, therefore, more easily testable. This makes it simpler to write unit tests and ensures that changes to one part of the codebase do not inadvertently break other components.
+
+## Design Pattern Examples
+
+Having introduced what a design patterns is discussed the various advantages design patterns offer to us as developers in our daily life. Lets take a look at the various design patterns that we usually use in conjujnction to Kotlin.
+
+## Builder Design Pattern
+
+The Builder design pattern is used for constructing complex objects by separating the construction process from the actual representation. It is particularly useful when an object has a large number of parameters, and we want to provide a more readable and flexible way to construct it.
+
+Here's an example of implementing the Builder design pattern in Kotlin:
+
+```kotlin
+// Product class
+data class Computer(
+    val cpu: String,
+    val ram: String,
+    val storage: String,
+    val graphicsCard: String
+)
+
+// Builder interface
+interface ComputerBuilder {
+    fun buildCpu(cpu: String): ComputerBuilder
+    fun buildRam(ram: String): ComputerBuilder
+    fun buildStorage(storage: String): ComputerBuilder
+    fun buildGraphicsCard(graphicsCard: String): ComputerBuilder
+    fun build(): Computer
+}
+
+// Concrete builder class
+class ConcreteComputerBuilder : ComputerBuilder {
+    private var cpu: String = ""
```

Comment:
This effectively kills Kotlin's null-safety feature. The builder allows the client to NOT set a certain attribute and it will still be the empty String after `build()` is being called. This is not possible when using named parameters ... another reason why named parameters should be the first choice, in my opinion.